Cluster Access Tooling Architecture Docs (#479)
appuio/openshift4-docs: docs/modules/ROOT/pages/references/architecture/cluster_access_tooling.adoc

=== CLI

==== `update`

The `update` sub-command updates the routing information by querying the Lieutenant API and caching the information locally.
See <<routing-information-retrieval>> for more information.

==== `proxy`

The `proxy` sub-command sets up a local SOCKS5 proxy that routes traffic to cluster endpoints through the appropriate jump hosts based on the routing information retrieved from the Lieutenant API.
All other traffic is directly connected to the destination.

The command fails fast if the routing information isn't available, so users are expected to run `update` before running `proxy` for the first time, and then run `update` periodically to keep the routing information up to date.

See <<socks5-proxy>> for more information.

==== `test`

The `test` sub-command tests the connectivity to all jump hosts and clusters based on the routing information retrieved from the Lieutenant API.
This can be used to verify that the routing information is correct and that the jump hosts and clusters are accessible.

==== `clusters`

The `clusters` sub-command lists the clusters based on the information retrieved from the Lieutenant API.

> ==== `clusters CLUSTER_NAME kubeconfig`

Reviewer: Might be interesting to also have a variant of this command which generates kubeconfigs for all known clusters (based on Lieutenant API?) or a single kubeconfig with a context for each cluster. I'm not picky about the exact format (personally I want separate kubeconfigs per cluster, but I already have https://github.com/simu/kubectl-split-config to generate per-cluster kubeconfigs from a big kubeconfig with many contexts)

Also, this probably needs more information than the facts currently outlined in this document (especially for clusters that don't need a jumphost).

The `clusters CLUSTER_NAME kubeconfig` sub-command retrieves the kubeconfig for the specified cluster and writes it to the specified file or standard output if no file is specified.

==== `clusters CLUSTER_NAME emergency-credentials`

> The `clusters CLUSTER_NAME emergency-credentials` sub-command retrieves the emergency credentials kubeconfig for the specified cluster and writes it to the specified file or standard output if no file is specified.

Reviewer: Same here: for seamless use, this command needs to know the API URL of each cluster -> we probably should add that as a fact (either as a fact during cluster setup, or through Steward's additional dynamic facts configmap, since we already have the API URL for OCP clusters in the inventory).

The Go SSH tooling allows to create SSH (agent) servers and clients and we easily end-to-end test the SSH tunneling functionality without needing to set up actual jump hosts and clusters.

A full POC has been implemented using the Go SSH tooling and was able to connect to all our managed clusters with up to three jumps through jump hosts.
All SSH features we use in our VSHN SSH configuration are supported, and the implementation feels stable, even when changing networks or under load, and performs well.

=== Routing information retrieval [[routing-information-retrieval]]

Routing information is queried from our central inventory, which is managed using https://syn.tools/syn/index.html[Project Syn] and can be queried through the https://syn.tools/lieutenant-api/references/index.html[Lieutenant API].

Clusters that need a jump host to access have an additional cluster fact that specifies the jump host to use for accessing the cluster.

[source,yaml]
----
apiVersion: syn.tools/v1alpha1
kind: Cluster
metadata:
  name: c-bettersmarter-prod01
spec:
  displayName: BetterSmarter Prod01
  facts:

>     apiURL: https://api.c-bettersmarter-prod01.vshnmanaged.net:6443 <1>

Author: Not sure if we want to make it a dynamic fact? It feels right in the cluster facts and feels bit bloated to add it with a rountrip through compile->argocd->steward. *But* we already have it in the catalog.

Reviewer: No hard feelings either way, I suggested the dynamic fact because that would reuse the current source of truth parameters for the cluster API URL etc.

If we go with static facts, We should also consider the following approach: add facts that mimic the current `openshift` parameters (and replace the default values for those with the appropriate facts):

Currently (global defaults):

```
  openshift:
    baseDomain: ${cluster:name}.example.com
    appsDomain: apps.${openshift:baseDomain}
    apiURL: api.${openshift:baseDomain}
```

After (Lieutenant):

```
facts:
  openshiftBaseDomain: c-bettersmarter-prod01.vshnmanaged.net
  openshiftAppsDomain: apps.${facts:openshiftBaseDomain}
  openshiftApiURL: api.${facts:openshiftBaseDomain}
```

and (global defaults)

```
openshift:
  baseDomain: ${facts:openshiftBaseDomain}
  appsDomain: ${facts:openshiftAppsDomain}
  apiURL: ${facts:openshiftApiURL}
```

I've quickly verified that using references in facts works fine in Commodore (would need special handling in other tools though). Overall, it's not super clean either, because we now need to set 3 additional facts for each new cluster, two of which usually will be left as default.

If we go with static facts, I'd omit the API scheme and port in the fact since it's always 6443 for OpenShift and if either are included in the fact, we can't use the fact to set parameter `openshift.apiURL` which is used in places where we don't want the port and scheme.

Reviewer: New idea from discussion with @HappyTetrahedron and @bastjan: go with dynamic facts for OpenShift base domain, apps domain, API URL and console URL. This doesn't require a new source of truth for the values (we keep the source of truth in inventory parameter `openshift`), doesn't require custom rendering of reclass references outside Commodore, and allows us to generate a default `jumphostDomains` set from the dynamic facts.

Additionally, having these dynamic facts will be generally useful for other integrations that consume data from the Lieutenant API (e.g. Gandalf workflows) or Lieutenant metrics (e.g. Grafana dashboards).

Roughly, we'd have something like the following:

```
status:
  facts:
    openshiftBaseDomain: c-bettersmarter-prod01.vshnmanaged.net
    openshiftAppsDomain: apps.c-bettersmarter-prod01.vshnmanaged.net
    openshiftApiURL: api.c-bettersmarter-prod01.vshnmanaged.net
    openshiftConsoleURL: console.apps.c-bettersmarter-prod01.vshnmanaged.net
```

The (naively) generated `jumphostDomains` would be `['c-bettersmarter-prod01.vshnmanaged.net', 'apps.c-bettersmarter-prod01.vshnmanaged.net', 'api.c-bettersmarter-prod01.vshnmanaged.net', 'console.apps.c-bettersmarter-prod01.vshnmanaged.net']`. While this has a few redundant entries, this accounts for any custom values for apps domain, API URL and console URL.

With this approach, the values in (probably now optional) static fact `jumphostDomains`/`jumphostAdditionalDomains` will be added to the generated list.

Not sure right now whether we'd include scheme and port for API and console URLs with this approach, since we probably don't want those for the actual proxy routing.

Author: > Not sure right now whether we'd include scheme and port for API and console URLs with this approach, since we probably don't want those for the actual proxy routing.

If we name it URL we should add them. This would also allow integrations if the ports ever change. It's rather easy in most languages to parse URLs.

    jumphost: management4.rma1.bettersmarter.vshnmanaged.net <2>
    jumphostDomains: c-bettersmarter-prod01.vshnmanaged.net,vcenter.bettersmarter.ch,storage.bettersmarter.ch <3>

>     noJumphostDomains: apps.c-bettersmarter-prod01.vshnmanaged.net <4>

Author: Better naming welcome 🥺

Reviewer: not sure if it's better, but we could mimic the Linux `no_proxy` naming:

```suggestion
    proxyDomains: c-bettersmarter-prod01.vshnmanaged.net,vcenter.bettersmarter.ch,storage.bettersmarter.ch <3>
    noProxyDomains: apps.c-bettersmarter-prod01.vshnmanaged.net <4>
```

Reviewer: based on discussion with @HappyTetrahedron and @bastjan 

```suggestion
    jumphostDomains: c-bettersmarter-prod01.vshnmanaged.net,vcenter.bettersmarter.ch,storage.bettersmarter.ch <3>
    jumphostSkipDomains: apps.c-bettersmarter-prod01.vshnmanaged.net <4>
```

or with the discussion above about generating a default `jumphostDomains`:

```suggestion
    jumphostAdditionalDomains: vcenter.bettersmarter.ch,storage.bettersmarter.ch <3>
    jumphostSkipDomains: apps.c-bettersmarter-prod01.vshnmanaged.net <4>
```

spec:
  displayName: BetterSmarter Prod01
  facts:
    apiURL: https://api.c-bettersmarter-prod01.vshnmanaged.net:6443 <1>
    jumphost: management4.rma1.bettersmarter.vshnmanaged.net <2>
    jumphostDomains: c-bettersmarter-prod01.vshnmanaged.net,vcenter.bettersmarter.ch,storage.bettersmarter.ch <3>
    noJumphostDomains: apps.c-bettersmarter-prod01.vshnmanaged.net <4>
status:
  facts:
    openshiftOAuthRoute: '"oauth-openshift.apps.c-bettersmarter-prod01.vshnmanaged.net"' <5>
----
<1> The API URL of the cluster, which is used for the generated kubeconfig.
<2> The hostname of the jump host to use for accessing the cluster.
Only the last host in the jump host chain should be specified, the tooling looks up the routing information for the jump host by querying the SSH configuration.
<3> Comma separated list of additional domains that should be routed through the same jump host.
This can be used for access to management interfaces such as the vCenter interface the cluster is running on.
<4> Comma separated list of domains that shouldn't be routed through the jump host.
This can be used for access to cluster endpoints that aren't accessible from the DMZ.

> <5> Already existing dynamic fact that can be used for `oc login`.

Reviewer: Afaik, `oc login` doesn't need the oauth route, you just point it at the cluster's API URL.

+
[source]
----
Host management4.rma1.bettersmarter.vshnmanaged.net
  ProxyJump vpn1.rma1.bettersmarter.vshnmanaged.net

Host vpn1.rma1.bettersmarter.vshnmanaged.net
  ProxyCommand ssh -W vpn1.rma1.bettersmarter.vshnmanaged.net:22 -- vpn5.corp.vshn.net
----

The tooling authenticates to the Lieutenant API using OIDC with the same flow as `commodore fetch-token`.

For ease of management, the proxy reloads the routing information and invalidates caches on file change.
Existing connections are however not torn down.

=== Running as a Daemon

The tooling doesn't implement any special demonization logic, we use standard process management tools to run the proxy in the background, such as `systemd` on Linux and `launchd` on macOS.

==== `systemd`

The tool should provide a `systemd` user service file that can be used to run the proxy in the background on Linux.

User services get the users SSH configuration and `ssh-agent`.
We tested this with Gnome based desktop environments (Fedora 43, Ubuntu 26.04).

.systemd user service file example
[%collapsible]
====
[source,ini]
....
# ~/.config/systemd/user/smart-access.service
[Unit]
Description=Smart Access cluster proxy

[Service]

> Environment="SSH_AUTH_SOCK=%t/gcr/ssh" <1>

Reviewer: ~~Most likely this will not work for everyone out of the box.~~

After some more testing, this might work on all systemd-based distros (including auto key unlock with PAM).

Author: I only tested on Gnome based systems to be honest.

Reviewer: Yes, after some more research, GCR is part of Gnome Cryptography services, so non-Gnome environments may work differently.

==== `clusters`

The `clusters` sub-command lists the clusters based on the information retrieved from the Lieutenant API.

==== `clusters CLUSTER_NAME kubeconfig`

The `clusters CLUSTER_NAME kubeconfig` sub-command retrieves the kubeconfig for the specified cluster and writes it to the specified file or standard output if no file is specified.

==== `clusters CLUSTER_NAME emergency-credentials`

The `clusters CLUSTER_NAME emergency-credentials` sub-command retrieves the emergency credentials kubeconfig for the specified cluster and writes it to the specified file or standard output if no file is specified.

See xref:oc4:ROOT:references/architecture/emergency_credentials.adoc[] for implementation details of the emergency credentials retrieval.

==== `clusters CLUSTER_NAME shell`

[NOTE]
----
We might try auto-login to the cluster using `oc login` before opening the shell.
----

The `clusters CLUSTER_NAME shell` sub-command opens a shell with the kubeconfig set up to access the specified cluster.

.Interactive shell with kubeconfig example
[%collapsible]
====
[source,go]
....
package main

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
)

func main() {

> 	kubeconfig := "/tmp/kubeconfig"

Reviewer: for a real version this should probably be a cluster-specific path, otherwise running multiple `clusters CLUSTER_NAME shell` for different clusters in parallel will create chaos.

Author: True, my intentions was always something like `$XDG_CACHE_HOME/tool/c-cluster1/kubeconfig`. Will add a callout.

= Cluster Access Tooling

[abstract]
--

> A proposal for a new cluster access tooling to provide a user-friendly solution to access our managed OpenShift clusters without a public API.

Reviewer: ```suggestion
This document proposes a new cluster access tooling which aims to provide a user-friendly solution to access our managed OpenShift clusters without a public API.
```

> The proposal includes a SOCKS5 proxy routing connections, and cli tooling.

Reviewer: ```suggestion
The proposal includes a SOCKS5 proxy routing connections, and CLI tooling.
```

It checks which OS it's running on and installs the appropriate service file.

==== `update`

The `update` sub-command updates the routing information by querying the Lieutenant API and caching the information locally.
See <<routing-information-retrieval>> for more information.

==== `proxy`

The `proxy` sub-command sets up a local SOCKS5 proxy that routes traffic to cluster endpoints through the appropriate jump hosts based on the routing information retrieved from the Lieutenant API.
All other traffic is directly connected to the destination.

The command fails fast if the routing information isn't available, so users are expected to run `update` before running `proxy` for the first time, and then run `update` periodically to keep the routing information up to date.

See <<socks5-proxy>> for more information.

==== `test`

The `test` sub-command tests the connectivity to all jump hosts and clusters based on the routing information retrieved from the Lieutenant API.
This can be used to verify that the routing information is correct and that the jump hosts and clusters are accessible.

==== `kubeconfig`

> The `kubeconfig` sub-command generate a kubeconfig for all known clusters based on the routing information retrieved from the Lieutenant API and writes it to the specified file or standard output if no file is specified.

Reviewer: Follow-up question here: will this cover all clusters or just all clusters that match `.facts.distribution=="openshift4" or .facts.distribution=="oke"`?

Or just clusters that actually have the `openshiftApiURL` dynamic fact?

Author: Added a note https://github.com/appuio/openshift4-docs/pull/479/changes#diff-49442fa71adf71163820140781121d06e0e60bde92e763f9507bf06264e62f98R336-R340

Missing or empty `jumphost` means the cluster is directly accessible and doesn't require a jump host.
<2> The `jumphostDomains` fact specifies domains that should be routed through the jump host.
Domains are comma-separated suffixes.
Domains from `.status.facts` (dynamic facts) are added to this list.
<3> The `jumphostSkipDomains` fact specifies domains that should be excluded from routing through the jump host, even if they match the `jumphostDomains` list.
Domains are comma-separated suffixes.
<4> Dynamic facts are retrieved from the cluster itself, such as the API URL and console URL.
Those facts are added to the list of domains to route through the jump host, so users don't have to manually add them to the `jumphostDomains` fact.
The domain suffix list generated from this cluster would look like this:
+
[source,yaml]
----
exclude:
- apps.c-bettersmarter-prod01.vshnmanaged.net
include:
- console.apps.c-bettersmarter-prod01.vshnmanaged.net
- apps.c-bettersmarter-prod01.vshnmanaged.net
- api.c-bettersmarter-prod01.vshnmanaged.net
- c-bettersmarter-prod01.vshnmanaged.net
- storage.bettersmarter.ch
- vcenter.bettersmarter.ch
----

> Information downloaded from the Lieutenant API is cached locally to allow the tooling to continue working even if the Lieutenant API is temporarily unavailable.

Reviewer 2: Will this be in a accessible location and format for editing temporarily? I sometimes find myself in a situation where I quickly want to edit `jumphostDomains`, i.e. edit the foxyproxy settings in our current setup.

Author: Yes most probably `XDG_CACHE_DIR` on linux.

It checks which OS it's running on and installs the appropriate service file.

==== `update`

The `update` sub-command updates the routing information by querying the Lieutenant API and caching the information locally.
See <<routing-information-retrieval>> for more information.

==== `proxy`

The `proxy` sub-command sets up a local SOCKS5 proxy that routes traffic to cluster endpoints through the appropriate jump hosts based on the routing information retrieved from the Lieutenant API.
All other traffic is directly connected to the destination.

The command fails fast if the routing information isn't available, so users are expected to run `update` before running `proxy` for the first time, and then run `update` periodically to keep the routing information up to date.

See <<socks5-proxy>> for more information.

==== `test`

The `test` sub-command tests the connectivity to all jump hosts and clusters based on the routing information retrieved from the Lieutenant API.
This can be used to verify that the routing information is correct and that the jump hosts and clusters are accessible.

==== `kubeconfig`

> The `kubeconfig` sub-command generate a kubeconfig for all known clusters based on the routing information retrieved from the Lieutenant API and writes it to the specified file or standard output if no file is specified.

Reviewer 2: I lack some basic understanding here; does creating a kubeconfig automatically imply a login to the cluster? Can I run kubectl commands immediately after running this or does it still require authentication first?

Reviewer: My understanding for `kubeconfig` and `clusters CLUSTER_NAME kubeconfig` is that the commands simply generate a kubeconfig file which you can then use by pointing `kubectl`/`oc` to it by setting the `KUBECONFIG` env variable.

If you want to directly do stuff, you're probably looking for the `clusters CLUSTER_NAME shell` command.

Author: > does creating a kubeconfig automatically imply a login

No, that's out of scope as there's many ways of logging into a kubernetes cluster.

I will implement an additional OCP specific command.

A JSON serialized https://kubernetes.io/docs/reference/config-api/kubeconfig.v1/#AuthInfo[kubeconfig.v1.AuthInfo] object can be provided as a cluster fact which the tooling then uses to generate the kubeconfig.
----

==== `clusters`

The `clusters` sub-command lists the clusters based on the information retrieved from the Lieutenant API.

==== `clusters CLUSTER_NAME kubeconfig`

The `clusters CLUSTER_NAME kubeconfig` sub-command retrieves the kubeconfig for the specified cluster and writes it to the specified file or standard output if no file is specified.

==== `clusters CLUSTER_NAME emergency-credentials`

The `clusters CLUSTER_NAME emergency-credentials` sub-command retrieves the emergency credentials kubeconfig for the specified cluster and writes it to the specified file or standard output if no file is specified.

See xref:oc4:ROOT:references/architecture/emergency_credentials.adoc[] for implementation details of the emergency credentials retrieval.

==== `clusters CLUSTER_NAME shell`

[NOTE]
----
We might try auto-login to the cluster using `oc login` before opening the shell.
----

> The `clusters CLUSTER_NAME shell` sub-command opens a shell with the kubeconfig set up to access the specified cluster.

Reviewer: Idea here: have an option to let the user specify a shell script which gets sourced before the shell is handed to the user -> allows doing stuff like customizing PS1 to the users preference (or run a suitable login command without having to bake that into the tool).

Author: Or at least an environment variable a `.bashrc` or similar could pick up on 👍
